julia/related.jl

```diff
@@ -10,7 +10,6 @@ function relatedIO()
     posts = JSON3.read(json_string, Vector{PostData})    
 
     start = now()
-    related(posts)
```

Comment:
If you remove the warmup call then we are arriving to something similar to this:

![image](https://github.com/jinyus/related_post_gen/assets/3482227/4d15ed8f-4d9e-4ca8-9c88-40d6079a5352)

This is properly [documented ](https://docs.julialang.org/en/v1/manual/performance-tips/#Measure-performance-with-[@time](@ref)-and-pay-attention-to-memory-allocation) here.

If you are stumbling upon the runtime dispatch of the unsigned int types, that has nothing to do with this.

Please keep in mind that Julia does not compile the code in the same manner as Go/Rust.

You can read more about this [here](https://docs.julialang.org/en/v1/manual/performance-tips/#Measure-performance-with-[@time](@ref)-and-pay-attention-to-memory-allocation).

Reply:
Alright, I understood that this is not the type assertions that this is used for but for warming up (compiling the function using the JIT compiler for later use). But are we sure that there is no caching happening? The type assertions for example have to be run during the measured time. But if the function is warmed up with real data, it may already cache some data.

Also: I started a discussion issue:
https://github.com/jinyus/related_post_gen/issues/205

Reply:
>But are we sure that there is no caching happening? 

the only thing get's cached is the machine code of the function. When you call it the next time, it still needs to check length of `posts` and do a runtime dynamic dispatch

Reply:
Julia's JIT is not like most traditional hotspot jits. It doesn't cache based on runtime values (unless one explicitly lifts those values to the type domain).

Reply:
So would it be fine to do the warmup without the actual data like this:
```
# warum up without the actual data
related(PostData[])
```

instead of 
`related(posts)`

Reply:
yes, it would absolutely be fine

Reply:
> yes, it would absolutely be fine

I think we need to test it. I am talking about some conditional function calls that can be impacted by the size of the data.

Just a silly example:
```julia
f() = "something that needs compiling"

function f1(xs)
    for (i, _) in enumerate(xs)
        if i > 25 # this number can be runtime dependend
            f()
        end
    end
end
```

If we call `f1` with something with a size less than 26, the `f` function might not be called (and compiled) at all.

Not saying this is actually the case with our current scenario - just saying that such "small" calls might work or might not.

Reply:
the single-argument `related()` is just a run-time dispatcher, it ends up calling one of:
```julia
related(UInt8, posts)
related(UInt16, posts)
related(UInt32, posts)
related(UInt64, posts)
```

they do the real work, if we force compile each of them, I don't see it going wrong

Reply:
you may also have a slightly off understanding of Julia compile model:
```julia
julia> @generated function f(x)
           println("this is compile time for $x")
           return 1
       end
f (generic function with 1 method)

julia> function f1(xs)
           for (i, _) in enumerate(xs)
               if i > 25 # this number can be runtime dependend
                   f(1)
               end
           end
       end
f1 (generic function with 1 method)

julia> f1([1])
this is compile time for Int64
```

Reply:
I checked it right now and it seems to be equivalent in speed to me

Reply:
```
for T in (UInt8, UInt16, UInt32, UInt64)
````
I assume someone has benchmarked this and verified that there truly is a performance improvement for narrower types? It seems that there isn't a lot of room for SIMD here so I wouldn't have expected much difference.

Reply:
SIMD is not allowed: https://github.com/jinyus/related_post_gen/issues/126
Also:
https://github.com/jinyus/related_post_gen/pull/166#issuecomment-1752179043